streaming/stream_transfer_task.cc

```diff
@@ -203,15 +204,27 @@ future<> send_mutation_fragments(lw_shared_ptr<send_info> si) {
         }();
 
         auto sink_op = [sink, si, got_error_from_peer] () mutable -> future<> {
-            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink) {
-                return repeat([&sink, si, got_error_from_peer] () mutable {
-                    return si->reader(db::no_timeout).then([&sink, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
-                        if (mf && !(*got_error_from_peer)) {
+            mutation_fragment_stream_validator validator(*(si->reader.schema()));
+            return do_with(std::move(sink), std::move(validator), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink, mutation_fragment_stream_validator& validator) {
+                return repeat([&sink, &validator, si, got_error_from_peer] () mutable {
+                    return si->reader(db::no_timeout).then([&sink, &validator, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
+                        if (*got_error_from_peer) {
```

Comment:
This is a poor API that requires you to check for an error, and if you forget, it corrupts data. Can we replace it with something that throws an exception?

Reply:
This is the limitation of the rpc. Assume n1 is the sender and n2 is the receiver. The stream (s1) n1 to n2 is used to send data. We need another stream (s2) to send the error from n2 to n1. 
The loop for s1 and s2 works inside different fibers. How do we use exception  in this case to signal the error between the fibers?

Reply:
Agree it's an rpc bug. I think we'll need to communicate errors on the main rpc connection, not via the streams. This will need an updated API. 

@gleb-cloudius